convert to dynamic config; return boolean, numeric and string responses (FF-860)

leoromanovsky opened the pull request:
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: #__issue__

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

Adds support for typed response functions.

## Description
[//]: # (Describe your changes in detail)

* enable strict null checks
* Keeps `getAssignment(...): string`
* Adds `getStringAssignment`, `getBoolAssignment`, `getNumericAssignment`, `getJSONAssignment`
* Created `EppoValue` class

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)

unit tests

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

leoromanovsky commented on src/client/eppo-client.ts:
```diff
@@ -32,46 +33,178 @@ export interface IEppoClient {
+      assignmentHooks,
+      ValueType.StringType,
+    );
+    assignmentHooks?.onPostAssignment(experimentKey, subjectKey, assignment);
 
+    if (assignment !== null)
+      this.logAssignment(experimentKey, assignment, subjectKey, subjectAttributes);
+
```
the pre and post assignment events caused a bunch of duplicate code in these handlers

agamkrbit commented on src/client/eppo-client.spec.ts:
```diff
@@ -15,6 +16,7 @@ import { IAssignmentLogger } from '../assignment-logger';
 import { IConfigurationStore } from '../configuration-store';
 import { MAX_EVENT_QUEUE_SIZE } from '../constants';
 import { OperatorType } from '../dto/rule-dto';
+import { EppoValue, IValue } from '../eppo_value';
```
Nit: please fix the lint error

agamkrbit commented on tsconfig.json:
```diff
@@ -8,6 +8,7 @@
     "target": "es2017",
     "outDir": "dist",
     "noImplicitAny": true,
+    "strictNullChecks": true,
```
👍

agamkrbit commented on src/dto/variation-dto.ts:
```diff
@@ -1,11 +1,12 @@
   name: string;
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  value: any;
+  typedValue: IValue;
```
`IValue` only supports `boolean`, `number` and `string`. But the typedValue can also be a json right?

leoromanovsky replied:
Yes in that case it will be a `string`; in this PR `IValue` is a type so that json deserialization works and `EppoValue` is the container for the various values.

What I am missing from being able to unify them is something like `UnmarshalJSON` in golang which is called when an object is being json deserialized.

https://github.com/Eppo-exp/golang-sdk/blob/main/eppoclient/value.go

Do you have an idea for how to accomplish this in javascript?

agamkrbit commented on src/client/eppo-client.ts:
```diff
@@ -124,14 +266,13 @@ export default class EppoClient implements IEppoClient {
 
   private logAssignment(
     experiment: string,
-    variation: string,
+    variation: EppoValue,
     subjectKey: string,
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    subjectAttributes: Record<string, any>,
+    subjectAttributes: Record<string, EppoValue> | undefined = {},
   ) {
     const event: IAssignmentEvent = {
       experiment,
-      variation,
+      variation: variation.stringValue ?? '', // should assignment event have EppoValue?
```
Good point, In Java I think I have kept variation as a string and returned string by converting all the types to string. If it is just for logging then string type is sufficient but if they are using it for some other purpose then send an EppoValue makes more sense.

aarsilv replied:
I think `.stringValue` is different than what @agamkrbit may be saying. You may want to implement a  `.toString()` override that returns the value as a string regardless of type. Here, you will return `undefined` unless the variation actually is a string type.

If you are going for a `.toString()` like behavior, I'd protect that with an automated test as well.

agamkrbit commented:
@leoromanovsky Can we test it with the updated test data from the GitHub Repository?

leoromanovsky commented:
Thanks Agam for the suggestion to add the new test cases the latest ones brought up a bug that I corrected.

I also had a hard time getting the new `test-data` command working as it is written in the java repo; seemed like a race condition was happening where `git clone` wasn't getting waited on and the script crashing trying to copy a json file. I've fixed it with the version as written:

```
testDataDir := test/data/
tempDir := ${testDataDir}temp/
gitDataDir := ${tempDir}sdk-test-data/
branchName := main
githubRepoLink := https://github.com/Eppo-exp/sdk-test-data.git
repoName := sdk-test-data
.PHONY: test-data
test-data: 
	rm -rf $(testDataDir)
	mkdir -p $(tempDir)
	git clone -b ${branchName} --depth 1 --single-branch ${githubRepoLink} ${gitDataDir}
	cp ${gitDataDir}rac-experiments-v3.json ${testDataDir}
	cp -r ${gitDataDir}assignment-v2 ${testDataDir}
	rm -rf ${tempDir}
```

aarsilv commented on Makefile:
```diff
@@ -25,12 +25,19 @@ help: Makefile
-	gsutil cp gs://sdk-test-data/rac-experiments-v3.json $(testDataDir)
-	gsutil cp -r gs://sdk-test-data/assignment-v2 $(testDataDir)
+	mkdir -p $(tempDir)
+	git clone -b ${branchName} --depth 1 --single-branch ${githubRepoLink} ${gitDataDir}
```
🔥

aarsilv commented on src/eppo_value.ts:
```diff
@@ -0,0 +1,39 @@
+  public stringValue: string | undefined;
+
+  constructor(
+    valueType: ValueType,
+    boolValue: boolean | undefined,
+    numericValue: number | undefined,
+    stringValue: string | undefined,
```
If you want to enforce exactly one of the values is provided, you could do that with union types.

Although really I think we want this constructor to be `private` so people have to use the static factory methods.

leoromanovsky replied:
Good suggestion thank you.

aarsilv commented on src/client/eppo-client.spec.ts:
```diff
@@ -281,9 +304,29 @@ describe('EppoClient E2E test', () => {
+      switch (valueTestType) {
+        case ValueTestType.BoolType: {
+          const ba = globalClient.getBoolAssignment(subjectKey, experiment);
+          if (ba === null) return null;
+          return EppoValue.Bool(ba);
```
minor but with short-circuiting and I think we could simplify this to `return ba && EppoValue.Bool(ba)`.

```node
> v = null
null
> v && Bool(v)
null
> v = true
true
> v && Boolean(v)
true
```

aarsilv commented on src/eppo_value.spec.ts:
```diff
@@ -0,0 +1,28 @@
+import { EppoValue } from './eppo_value';
+
+describe('EppoValue toString function', () => {
```
🙌
